File: backend/ts_template/devdonalds.ts
import express, { Request, Response } from "express";

// ==== Type Definitions, feel free to add or modify ==========================
interface cookbookEntry {
  name: string;
  type: string;
}

interface requiredItem {
  name: string;
  quantity: number;
}

interface recipe extends cookbookEntry {
  requiredItems: requiredItem[];
}

interface ingredient extends cookbookEntry {
  cookTime: number;
}

// =============================================================================
// ==== HTTP Endpoint Stubs ====================================================
// =============================================================================
const app = express();
app.use(express.json());

// Store your recipes here!
const cookbook: Record<string, recipe | ingredient> = {};

// Task 1 helper (don't touch)
app.post("/parse", (req:Request, res:Response) => {
  const { input } = req.body;

  const parsed_string = parse_handwriting(input)
  if (parsed_string == null) {
    res.status(400).send("this string is cooked");
    return;
  } 
  res.json({ msg: parsed_string });
  return;
  
});

// [TASK 1] ====================================================================
// Takes in a recipeName and returns it in a form that 
const parse_handwriting = (recipeName: string): string | null => {
  const cleanedName = recipeName
    .replace(/[-_]+/g, " ")
    .replace(/[^a-zA-Z ]/g, "")
    .trim();
  
  if (!cleanedName) return null;
  
  return cleanedName
    .split(" ")
    .map(word => word.charAt(0).toUpperCase() + word.slice(1).toLowerCase())
    .join(" ");
};


// [TASK 2] ====================================================================
// Endpoint that adds a CookbookEntry to your magical cookbook
app.post("/entry", (req:Request, res:Response) => {
  const { name, type, requiredItems, cookTime } = req.body;
  
  if (!name || !type) {
    return res.status(400).json({ error: "Missing 'name' or 'type'" });
  }

  if (type !== "recipe" && type !== "ingredient") {
    return res.status(400).json({ error: "Invalid type. Must be 'recipe' or 'ingredient'" });
  }

  if (cookbook[name]) {
    return res.status(400).json({ error: "Entry name must be unique" });
  }

  if (type === "recipe") {
    if (!Array.isArray(requiredItems)) {
      return res.status(400).json({ error: "Recipe must have a list of 'requiredItems'" });
    }
    
    const seenItems = new Set();
    for (const item of requiredItems) {
      if (!item.name || typeof item.quantity !== "number") {
        return res.status(400).json({ error: "Each requiredItem must have 'name' and 'quantity'" });
      }
      if (seenItems.has(item.name)) {
        return res.status(400).json({ error: "Duplicate requiredItem names are not allowed" });
      }
      seenItems.add(item.name);
    }
    
    cookbook[name] = { name, type, requiredItems };
  } else if (type === "ingredient") {
    if (typeof cookTime !== "number" || cookTime < 0) {
      return res.status(400).json({ error: "Ingredient must have a 'cookTime' >= 0" });
    }
    
    cookbook[name] = { name, type, cookTime };
  }

  return res.sendStatus(200);

});

const calculateRecipeSummary = (recipeName: string): { cookTime: number; ingredients: requiredItem[] } => {
  const recipe = cookbook[recipeName] as recipe;
  let totalCookTime = 0;
  const ingredientMap: Record<string, number> = {};

  for (const item of recipe.requiredItems) {
    const entry = cookbook[item.name];
    if (!entry) {
      throw new Error(`Missing ingredient: ${item.name}`);
    }

    if (entry.type === "ingredient") {
      const ingredient = entry as ingredient;
      totalCookTime += ingredient.cookTime * item.quantity;
      ingredientMap[item.name] = (ingredientMap[item.name] || 0) + item.quantity;
    } else {
      const { cookTime, ingredients } = calculateRecipeSummary(item.name);
      totalCookTime += cookTime * item.quantity;
      for (const ingr of ingredients) {
        ingredientMap[ingr.name] = (ingredientMap[ingr.name] || 0) + ingr.quantity * item.quantity;
      }
    }
  }

  const ingredientsList = Object.entries(ingredientMap).map(([name, quantity]) => ({ name, quantity }));
  return { cookTime: totalCookTime, ingredients: ingredientsList };
};

// [TASK 3] ====================================================================
// Endpoint that returns a summary of a recipe that corresponds to a query name
app.get("/summary", (req:Request, res:Request) => {
  const name = req.query.name as string;
  if (!name || !cookbook[name]) {
    return res.status(400).json({ error: "Recipe not found" });
  }

  const recipe = cookbook[name];
  if (recipe.type !== "recipe") {
    return res.status(400).json({ error: "Requested name is not a recipe" });
  }

  try {
    const { cookTime, ingredients } = calculateRecipeSummary(name);
    return res.json({ name, cookTime, ingredients });
  } catch (error) {
    return res.status(400).json({ error: error.message });
  }

});

// =============================================================================
// ==== DO NOT TOUCH ===========================================================
// =============================================================================
const port = 8080;
app.listen(port, () => {
  console.log(`Running on: http://127.0.0.1:8080`);
});

File: backend/py_template/devdonalds.py
from dataclasses import dataclass
from typing import List, Dict, Union
from flask import Flask, request, jsonify
import re

# ==== Type Definitions, feel free to add or modify ===========================
@dataclass
class CookbookEntry:
	name: str

@dataclass
class RequiredItem():
	name: str
	quantity: int

@dataclass
class Recipe(CookbookEntry):
	required_items: List[RequiredItem]

@dataclass
class Ingredient(CookbookEntry):
	cook_time: int


# =============================================================================
# ==== HTTP Endpoint Stubs ====================================================
# =============================================================================
app = Flask(__name__)

# Put this to avoid 404 error
@app.route("/")
def home():
    return "Hello, DevDonalds!"

# Store your recipes here!
cookbook = {}

# Task 1 helper (don't touch)
@app.route("/parse", methods=['POST'])
def parse():
	data = request.get_json()
	recipe_name = data.get('input', '')
	parsed_name = parse_handwriting(recipe_name)
	if parsed_name is None:
		return 'Invalid recipe name', 400
	return jsonify({'msg': parsed_name}), 200

# [TASK 1] ====================================================================
# Takes in a recipeName and returns it in a form that 
def parse_handwriting(recipeName: str) -> Union[str | None]:
	recipeName = re.sub(r"[-_]+", " ", recipeName)

	recipeName = re.sub(r"[^a-zA-Z ]", "", recipeName)

	words = recipeName.split()

	if not words:
		return None
	
	return " ".join(word.capitalize() for word in words)


# [TASK 2] ====================================================================
# Endpoint that adds a CookbookEntry to your magical cookbook
@app.route('/entry', methods=['POST'])
def create_entry():
	data = request.get_json()

	if "name" not in data or "type" not in data:
		return jsonify({"error": "Missing 'name' or 'type'"}), 400
	
	name = data["name"]
	entry_type = data["type"]

	if entry_type not in {"recipe", "ingredient"}:
		return jsonify({"error": "Invalid type. Must be 'recipe' or 'ingredient'"}), 400
	
	if name in cookbook:
		return jsonify({"error": "Entry name must be unique"}), 400
	
	if entry_type == "recipe":
		if "requiredItems" not in data or not isinstance(data["requiredItems"], list):
			return jsonify({"error": "Recipe must have a list of 'requiredItems"}), 400
		
		required_items = data["requiredItems"]
		seen_items = set()

		for item in required_items:
			if "name" not in item or "quantity" not in item:
				return jsonify({"error": "Each requiredItem must have 'name' and 'quantity'"}), 400
			
			if item["name"] in seen_items:
				return jsonify({"error": "Duplicate requiredItem names are not allowed"}), 400
			
			seen_items.add(item["name"])
		
		cookbook[name] = data
	
	elif entry_type == "ingredient":
		if "cookTime" not in data or not isinstance(data["cookTime"], int) or data["cookTime"] < 0:
			return jsonify({"error": "Ingredient must have a 'cookTime' >= 0"}), 400
		
		cookbook[name] = data

	return "", 200

# [TASK 3] ====================================================================
# Endpoint that returns a summary of a recipe that corresponds to a query name
@app.route('/summary', methods=['GET'])
def summary():
	# TODO: implement me
	return 'not implemented', 500


# =============================================================================
# ==== DO NOT TOUCH ===========================================================
# =============================================================================

if __name__ == '__main__':
	app.run(debug=True, port=8080)
